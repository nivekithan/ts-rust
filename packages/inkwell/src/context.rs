use llvm_sys::{core::{LLVMAppendBasicBlockInContext, LLVMContextCreate, LLVMContextDispose, LLVMCreateBuilderInContext, LLVMDoubleTypeInContext, LLVMInt1TypeInContext, LLVMInt64TypeInContext, LLVMModuleCreateWithNameInContext, LLVMVoidTypeInContext}, prelude::LLVMContextRef};

use crate::{
    basic_block::BasicBlock,
    builder::Builder,
    module::Module,
    types::{float_type::FloatType, int_type::IntType, void_type::VoidType},
    utils::to_c_str,
    values::{fn_value::FunctionValue, traits::AsValueRef},
};

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Context {
    pub(crate) context: LLVMContextRef,
}

impl Context {
    pub(crate) unsafe fn new(context: LLVMContextRef) -> Self {
        assert!(!context.is_null());

        return Context { context };
    }

    pub fn create() -> Self {
        unsafe { return Context::new(LLVMContextCreate()) }
    }

    pub fn create_module(&self, name: &str) -> Module {
        let c_string = to_c_str(name);

        unsafe {
            return Module::new(LLVMModuleCreateWithNameInContext(
                c_string.as_ptr(),
                self.context,
            ));
        }
    }

    pub fn i1_type(&self) -> IntType {
        unsafe {
            return IntType::new(LLVMInt1TypeInContext(self.context));
        }
    }

    pub fn i64_type(&self) -> IntType {
        unsafe {
            return IntType::new(LLVMInt64TypeInContext(self.context));
        }
    }

    pub fn f64_type(&self) -> FloatType {
        unsafe { return FloatType::new(LLVMDoubleTypeInContext(self.context)) }
    }

    pub fn void_type(&self) -> VoidType {
        unsafe { return VoidType::new(LLVMVoidTypeInContext(self.context)) }
    }

    pub fn append_basic_block(&self, function: FunctionValue, name: &str) -> BasicBlock {
        let c_name = to_c_str(name);

        unsafe {
            return BasicBlock::new(LLVMAppendBasicBlockInContext(
                self.context,
                function.as_value_ref(),
                c_name.as_ptr(),
            ));
        }
    }

    pub fn create_builder(&self) -> Builder {
        unsafe {
            return Builder::new(LLVMCreateBuilderInContext(self.context));
        }
    }
}

impl Drop for Context {
    fn drop(&mut self) {
        unsafe {
            LLVMContextDispose(self.context);
        }
    }
}
