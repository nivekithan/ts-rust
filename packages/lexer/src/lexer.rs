use crate::token::LiteralKind;

use super::token::KeywordKind;
use core::panic;
use std::str::Chars;

use super::token::Token;

pub(crate) struct Lexer<'a> {
    content: Chars<'a>,
    cur_char: Option<char>,
}

impl<'a> Lexer<'a> {
    pub(crate) fn new(input: &'a str) -> Lexer<'a> {
        let content = input.chars();

        let mut lexer = Lexer {
            content,
            cur_char: None,
        };

        lexer.next();

        return lexer;
    }

    pub(crate) fn next_token(&mut self) -> Token {
        self.eat_whitespace();

        let cur_char = self.cur_char;

        match cur_char {
            None => Token::Eof,
            Some(char) => {
                use Token::*;

                if char == ';' {
                    self.next(); // consumes ;
                    return SemiColon;
                } else if char == ':' {
                    self.next(); // consumes :
                    return Colon;
                } else if char == '=' {
                    self.next(); // consumes =
                    return Assign;
                } else if char == '{' {
                    self.next(); // consumes {
                    return AngleOpenBracket;
                } else if char == '}' {
                    self.next(); // consumes }
                    return AngleCloseBracket;
                } else if char == '(' {
                    self.next(); // consumes (
                    return CurveOpenBracket;
                } else if char == ')' {
                    self.next(); // consumes )
                    return CurveCloseBracket;
                } else if char == ',' {
                    self.next(); // consumes ,
                    return Comma;
                } else if char == '!' {
                    self.next(); // consumes !
                    return Bang;
                } else if char == '+' {
                    self.next(); // consumes +
                    return Plus;
                } else if char == '-' {
                    self.next(); // consumes -
                    return Minus;
                } else if char == '*' {
                    self.next(); // consumes +
                    return Star;
                } else if char == '/' {
                    self.next(); // consumes /
                    return Slash;
                } else if char == '|' {
                    self.next(); // consumes |
                    return VerticalBar;
                } else if char == '^' {
                    self.next(); // consumes ^
                    return Caret;
                } else if char == '&' {
                    self.next(); // consumes &
                    return Ampersand;
                } else if char == '\'' {
                    let string_name = self.read_string('\'');
                    return Literal(LiteralKind::String { name: string_name });
                } else if char == '\"' {
                    let string_name = self.read_string('\"');
                    return Literal(LiteralKind::String { name: string_name });
                } else if char == '`' {
                    let string_name = self.read_string('`');
                    return Literal(LiteralKind::String { name: string_name });
                } else {
                    if is_letter(&char) {
                        let ident_name = self.read_identifier();
                        let may_be_keyword = is_keyword(&ident_name);

                        match may_be_keyword {
                            IsKeyword::Yes(kind) => return Keyword(kind),
                            IsKeyword::No => return Ident { name: ident_name },
                        }
                    } else if is_digit(&char) {
                        let digit_name = self.read_digit();
                        let digit_value: f64 = digit_name.parse().unwrap();

                        return Literal(LiteralKind::Float {
                            name: digit_name,
                            value: digit_value,
                        });
                    }
                    self.next(); // consumes Illegal
                    return Illegal;
                }
            }
        }
    }

    fn next(&mut self) -> Option<char> {
        let next_cur = self.content.next();
        self.cur_char = next_cur;

        return next_cur;
    }

    // If cur_char is whitespace it wil eat char until cur_char is not whitespace
    fn eat_whitespace(&mut self) {
        let cur_char = self.cur_char;

        if let Some(mut c) = cur_char {
            while is_whitespace(&c) {
                let next_char = self.next();
                if let Some(ch) = next_char {
                    c = ch
                } else {
                    break;
                }
            }
        }
    }

    // After executing this function the cur_char wont be a letter
    fn read_identifier(&mut self) -> String {
        let mut ident_name = self.cur_char.expect("unreachable").to_string();
        let next_cur = self.next();

        match next_cur {
            None => return ident_name,
            Some(mut c) => {
                while is_letter(&c) {
                    ident_name.push(c);

                    let next_cur = self.next();

                    match next_cur {
                        None => return ident_name,
                        Some(ch) => c = ch,
                    }
                }

                return ident_name;
            }
        }
    }
    // After executing this function the cur_char wont be a digit
    fn read_digit(&mut self) -> String {
        let mut digit_name = self.cur_char.expect("unreachable").to_string();

        let next_cur = self.next();

        match next_cur {
            None => return digit_name,
            Some(mut c) => {
                while is_digit(&c) {
                    digit_name.push(c);

                    let next_cur = self.next();

                    match next_cur {
                        None => return digit_name,
                        Some(ch) => c = ch,
                    }
                }

                return digit_name;
            }
        }
    }
    // Assumes the cur_char is starting char of string_literal
    // Ex: ' " `
    // It will end after consuming the end char that
    // cur_char wont be end_char
    fn read_string(&mut self, end_char: char) -> String {
        let mut string_name = String::new();
        loop {
            let next_char = self.next();

            match next_char {
                None => panic!("Lexer error expected {} before end of file", end_char),
                Some(ch) => {
                    if ch == end_char {
                        self.next(); // consumes end_char
                        break;
                    } else {
                        string_name.push(ch)
                    }
                }
            }
        }

        return string_name;
    }
}

// From rust source code
fn is_whitespace(c: &char) -> bool {
    return matches!(
        c,
        '\u{0009}'   // \t
        | '\u{000A}' // \n
        | '\u{000B}' // vertical tab
        | '\u{000C}' // form feed
        | '\u{000D}' // \r
        | '\u{0020}' // space

        // NEXT LINE from latin1
        | '\u{0085}'

        // Bidi markers
        | '\u{200E}' // LEFT-TO-RIGHT MARK
        | '\u{200F}' // RIGHT-TO-LEFT MARK

        // Dedicated whitespace characters from Unicode
        | '\u{2028}' // LINE SEPARATOR
        | '\u{2029}' // PARAGRAPH SEPARATOR)
    );
}

fn is_letter(c: &char) -> bool {
    return matches!(
        c,
         'a'..='z'
        | 'A'..='Z'
        | '_'
    );
}

fn is_digit(c: &char) -> bool {
    return matches!(c, '0'..='9');
}

pub enum IsKeyword {
    Yes(KeywordKind),
    No,
}

fn is_keyword(word: &String) -> IsKeyword {
    if word == "const" {
        return IsKeyword::Yes(KeywordKind::Const);
    } else if word == "true" {
        return IsKeyword::Yes(KeywordKind::True);
    } else if word == "false" {
        return IsKeyword::Yes(KeywordKind::False);
    } else if word == "let" {
        return IsKeyword::Yes(KeywordKind::Let);
    } else if word == "if" {
        return IsKeyword::Yes(KeywordKind::If);
    } else {
        return IsKeyword::No;
    }
}
