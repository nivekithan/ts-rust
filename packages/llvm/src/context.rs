use llvm_sys::{
    core::{LLVMContextCreate, LLVMContextDispose},
    prelude::*,
};

use crate::{
    basic_block::BasicBlock, builder::Builder, module::Module, values::double::LlvmDouble,
};

pub struct Context {
    llvm_ref: LLVMContextRef,
}

impl Context {
    pub fn new() -> Context {
        unsafe {
            let llvm_ref = LLVMContextCreate();
            return Context { llvm_ref };
        }
    }

    pub(crate) fn get_ref(&self) -> LLVMContextRef {
        return self.llvm_ref;
    }

    // Creates a module in this context
    pub fn create_module(&self, name: &str) -> Module {
        return Module::new(self, name);
    }

    pub fn create_basic_block(&self, name: &str) -> BasicBlock {
        return BasicBlock::new(self, name);
    }

    pub fn create_builder(&self) -> Builder {
        return Builder::new(self);
    }

    pub fn create_double(&self, num: f64) -> LlvmDouble {
        return LlvmDouble::new(self, num);
    }
}

impl Drop for Context {
    fn drop(&mut self) {
        unsafe {
            LLVMContextDispose(self.llvm_ref);
        }
    }
}
