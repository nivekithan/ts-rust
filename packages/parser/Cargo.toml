[package]
name = "parser"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lexer = { path = "../lexer"}
ast = {path = "../ast"}
indexmap = "1.7.0"
