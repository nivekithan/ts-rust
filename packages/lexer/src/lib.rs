mod lexer;
mod token;

use crate::lexer::Lexer;
use crate::token::Token;

#[cfg(test)]
mod test;

pub fn convert_to_token(input: &str) -> Vec<Token> {
    let mut lexer = Lexer::new(input);
    let mut tokens: Vec<Token> = Vec::new();

    loop {
        let token = lexer.next_token();

        tokens.push(token);

        if let Token::Eof = token {
            break;
        }
    }

    return tokens;
}
