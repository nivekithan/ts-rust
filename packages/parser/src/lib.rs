mod parse_block;
mod parse_expression;
mod parse_type_expression;
mod parser;
mod symbol_table;
mod utils;

#[cfg(test)]
mod tests;

use crate::{parser::Parser, symbol_table::SymbolContext};
use ast::Ast;
use lexer::token::Token;

pub fn convert_to_ast(input: Vec<Token>) -> Vec<Ast> {
    let mut parser = Parser::new(&input);
    let mut asts: Vec<Ast> = vec![];
    let mut context = SymbolContext::new_empty_context();

    while parser.get_cur_token().unwrap() != &Token::Eof {
        let next_ast = parser.next_ast(&mut context);
        asts.push(next_ast);
    }

    return asts;
}

#[cfg(test)]
mod test {
    use std::collections::HashMap;

    use ast::{
        data_type::DataType,
        declaration::{VariableDeclarationKind},
        expression::Expression,
        Ast,
    };
    use indexmap::indexmap;
    use lexer::convert_to_token;

    use crate::convert_to_ast;

    #[test]
    fn test_2() {
        let input = "
        const x = {a : 1, b : 2};
        const y = x.a;";

        let mut exp_hash_map: HashMap<String, Expression> = HashMap::new();

        exp_hash_map.insert(
            "a".to_string(),
            Expression::FloatLiteralExp {
                name: "1".to_string(),
                value: 1.0,
            },
        );
        exp_hash_map.insert(
            "b".to_string(),
            Expression::FloatLiteralExp {
                name: "2".to_string(),
                value: 2.0,
            },
        );

        let expected_output: Vec<Ast> = vec![
            Ast::new_variable_declaration(
                "x_",
                Expression::ObjectLiteral {
                    expression: exp_hash_map,
                    data_type: DataType::ObjectType {
                        entries: indexmap! {"a".to_string() => DataType::Float, "b".to_string() => DataType::Float},
                    },
                },
                VariableDeclarationKind::Const,
            ),
            Ast::new_variable_declaration(
                "y_",
                Expression::DotMemberAccess {
                    container: Box::new(Expression::IdentExp {
                        data_type: DataType::ObjectType {
                            entries: indexmap! {"a".to_string() => DataType::Float, "b".to_string() => DataType::Float},
                        },
                        name: "x_".to_string(),
                    }),
                    argument: "a".to_string(),
                },
                VariableDeclarationKind::Const,
            ),
        ];

        let actual_output = convert_to_ast(convert_to_token(input));

        assert_eq!(expected_output, actual_output);
    }

    #[test]
    fn test_3() {
        let input = "
        function foo(x : number) : number {
            const x = 1;
        }";

        let expected_output = vec![Ast::new_function_declaration(
            indexmap! {"x".to_string() => DataType::Float},
            Box::new(vec![Ast::new_variable_declaration(
                "x_",
                Expression::FloatLiteralExp {
                    name: "1".to_string(),
                    value: 1.0,
                },
                VariableDeclarationKind::Const,
            )]),
            "foo_".to_string(),
            DataType::Float,
        )];

        let actual_output = convert_to_ast(convert_to_token(input));

        assert_eq!(expected_output, actual_output);
    }
}
