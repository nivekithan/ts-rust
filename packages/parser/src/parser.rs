use core::panic;

use ast::{
    data_type::DataType,
    declaration::{VariableAssignmentOperator, VariableDeclarationKind},
    expression::{BinaryOperator, Expression, UnaryOperator},
    Ast,
};
use lexer::token::{KeywordKind, LiteralKind, Token};

use crate::symbol_table::{SymbolContext, SymbolMetaInsert};

pub struct Parser<'a> {
    content: &'a Vec<Token>,
    cur_pos: Option<usize>,
}

impl<'a> Parser<'a> {
    pub fn new(content: &'a Vec<Token>) -> Parser<'a> {
        let mut parser = Parser {
            content,
            cur_pos: None,
        };

        parser.next();

        return parser;
    }

    pub fn next_ast(&mut self, global_context: &mut SymbolContext) -> Ast {
        return self.next_ast_in_context(global_context).unwrap();
    }


    fn next_ast_in_context(&mut self, context: &mut SymbolContext) -> Result<Ast, String> {
        let first_token = self.get_cur_token()?;
        let suffix = &context.suffix.clone();

        match first_token {
            Token::Keyword(keyword_kind) => match keyword_kind {
                KeywordKind::Const | KeywordKind::Let => {
                    let is_const = match keyword_kind {
                        KeywordKind::Const => true,
                        KeywordKind::Let => false,

                        _ => unreachable!(),
                    };

                    let kind = match keyword_kind {
                        KeywordKind::Const => VariableDeclarationKind::Const,
                        KeywordKind::Let => VariableDeclarationKind::Let,

                        _ => unreachable!(),
                    };

                    // let  name = format!("{}{}", self.next().get_ident_name()?.clone(), suffix); // consumes Const
                    let name = self.next().get_ident_name()?.clone(); // consumes const

                    self.next(); // consumes ident

                    let expected_data_type = match self.get_cur_token()? {
                        Token::Colon => {
                            self.next(); // consumes :
                            self.parse_type_declaration()
                        }

                        _ => DataType::Unknown,
                    };

                    self.assert_cur_token(&Token::Assign)?;

                    self.next(); // consumes =

                    let expression = self.parse_expression(1, context)?;

                    let expression_data_type = expression.get_data_type();

                    if expected_data_type != DataType::Unknown
                        && expected_data_type != expression_data_type
                    {
                        return Err(format!(
                            "Expected data type {:?} but got {:?}",
                            expected_data_type, expression_data_type
                        ));
                    }

                    let sym_meta = SymbolMetaInsert::create(expression_data_type, is_const);

                    if let Err(_) = context.insert(name.as_str(), sym_meta) {
                        return Err(format!(
                            "You cannot declare variable {} which is already declared",
                            name
                        ));
                    }

                    self.skip_semicolon()?;

                    let name_with_suffix = format!("{}{}", name, suffix);
                    return Ok(Ast::new_variable_declaration(
                        name_with_suffix.as_str(),
                        expression,
                        kind,
                    ));
                }

                KeywordKind::If => {
                    // let mut child_context = context.create_child_context();
                    let ast = self.parser_if_block(context)?;
                    return Ok(ast);
                }

                _ => {
                    return Err(format!(
                        "Update function next_ast\n Unexpected keyword, {:?}",
                        keyword_kind
                    ))
                }
            },

            Token::Ident { name } => {
                // let name = format!("{}{}",name.clone(), suffix);
                let name = name.clone();

                if let Some(sym_meta) = context.get(&name) {
                    if sym_meta.is_const {
                        return Err(format!("Cannot reassign a const variable"));
                    }

                    let data_type = sym_meta.data_type.clone();

                    self.next(); // consumes the ident

                    let operator = match self.get_cur_token()? {
                        Token::Assign => VariableAssignmentOperator::Assign,
                        Token::PlusAssign => VariableAssignmentOperator::PlusAssign,
                        Token::MinusAssign => VariableAssignmentOperator::MinusAssign,
                        Token::StarAssign => VariableAssignmentOperator::StarAssign,
                        Token::SlashAssign => VariableAssignmentOperator::SlashAssign,

                        tok => return Err(format!("Expected either one of the =, +=, -=, *=, /= assignment operators but got {:?}", tok)),
                    };

                    self.next(); // consumes =

                    let expression = self.parse_expression(1, context)?;

                    if expression.get_data_type() != data_type {
                        return Err(format!(
                            "Reassigning datatype {:?} to variable whose datatype is {:?}",
                            expression.get_data_type(),
                            data_type
                        ));
                    }

                    self.skip_semicolon()?;

                    let suffix_name = format!("{}{}", name, sym_meta.suffix);

                    return Ok(Ast::new_variable_assignment(
                        suffix_name.as_str(),
                        operator,
                        expression,
                    ));
                } else {
                    return Err(format!("Unknown variable {}", name));
                }
            }

            tok => return Err(format!("Unknown token: {:?}", tok)),
        }
    }

    fn parser_if_block(&mut self, context: &mut SymbolContext) -> Result<Ast, String> {
        let first_token = self.get_cur_token().unwrap();

        match first_token {
            Token::Keyword(KeywordKind::If) => {
                self.next(); // consumes if

                self.assert_cur_token(&Token::CurveOpenBracket)?;
                self.next(); // consumes (

                let condition = self.parse_expression(1, context)?;

                self.assert_cur_token(&Token::CurveCloseBracket)?;
                self.next(); // consumes )

                self.assert_cur_token(&Token::AngleOpenBracket)?;
                self.next(); // consumes {

                let cur_value = context.counter;
                let suffix = format!("{}{}", context.suffix, cur_value);
                context.counter += 1;

                let mut child_context = context.create_child_context(suffix);

                let mut if_block_ast : Vec<Ast> = vec![];

                while self.get_cur_token().unwrap() != &Token::AngleCloseBracket {
                    let ast = self.next_ast_in_context(&mut child_context)?;
                    if_block_ast.push(ast);
                }

                self.next(); // consumes }

                return Ok(Ast::new_if_block(condition, if_block_ast));

            }

            _ => panic!(
                "Expected parser_if_block to be called only when the cur_token is of Keyword if"
            ),
        }
    }

    fn next(&mut self) -> &Token {
        match self.cur_pos {
            None => {
                self.cur_pos = Some(0);
                return &self.content[0];
            }

            Some(value) => {
                if value >= self.content.len() - 1 {
                    panic!("cur_pos is at maximum value")
                }

                let next_value = value + 1;
                self.cur_pos = Some(next_value);
                return &self.content[next_value];
            }
        }
    }

    fn assert_cur_token(&self, token_type: &Token) -> Result<(), String> {
        let cur_token = self.get_cur_token()?;

        if cur_token != token_type {
            return Err(format!(
                "Expected token type to be {:?} but got {:?}",
                token_type, cur_token
            ));
        }

        return Ok(());
    }

    fn skip_semicolon(&mut self) -> Result<(), String> {
        let cur_token = self.get_cur_token()?;

        if let &Token::SemiColon = cur_token {
            self.next();
        }

        return Ok(());
    }

    pub fn get_cur_token(&self) -> Result<&Token, String> {
        if let Some(size) = self.cur_pos {
            return Ok(&self.content[size]);
        } else {
            return Err(
                "Next method should be called atleast one time before calling get_cur_token "
                    .to_string(),
            );
        }
    }

    fn parse_expression(
        &mut self,
        precedence: usize,
        context: &SymbolContext,
    ) -> Result<Expression, String> {
        let mut prefix_fun = self.get_prefix_exp(context)?;
        let next_token = self.get_cur_token()?.clone();

        while next_token != Token::SemiColon
            && precedence < Parser::get_non_prefix_precedence(&next_token)
        {
            let infix_fun = self.get_non_prefix_exp(prefix_fun, context)?;

            match infix_fun {
                Ok(exp) => {
                    prefix_fun = exp;
                }
                Err(exp) => {
                    prefix_fun = exp;
                    break;
                }
            }
        }

        return Ok(prefix_fun);
    }

    fn get_prefix_exp(&mut self, context: &SymbolContext) -> Result<Expression, String> {
        let cur_token = self.get_cur_token().unwrap();

        match cur_token {
            Token::Plus => {
                let precedence = Parser::get_prefix_precedence(&Token::Plus);

                self.next(); // consumes +
                let arg_exp = self.parse_expression(precedence, context)?;
                return Ok(Expression::UnaryExp {
                    operator: UnaryOperator::Plus,
                    argument: Box::new(arg_exp),
                });
            }

            Token::Minus => {
                let precedence = Parser::get_prefix_precedence(&Token::Minus);

                self.next(); // consumes -
                let arg_exp = self.parse_expression(precedence, context)?;
                return Ok(Expression::UnaryExp {
                    operator: UnaryOperator::Minus,
                    argument: Box::new(arg_exp),
                });
            }

            Token::Bang => {
                let precedence = Parser::get_prefix_precedence(&Token::Bang);

                self.next(); //  consumes !
                let arg_exp = self.parse_expression(precedence, context)?;
                return Ok(Expression::UnaryExp {
                    operator: UnaryOperator::Bang,
                    argument: Box::new(arg_exp),
                });
            }

            Token::Literal(literal_kind) => match literal_kind {
                LiteralKind::Float { name, value } => {
                    let name = name.to_string();
                    let value = *value;

                    self.next(); // consumes Float

                    return Ok(Expression::FloatLiteralExp { name, value });
                }

                LiteralKind::String { name } => {
                    let name = name.to_string();

                    self.next(); // consumes string

                    return Ok(Expression::StringLiteralExp { value: name });
                }
            },

            Token::Keyword(keyword_kind) => match keyword_kind {
                KeywordKind::True => {
                    self.next(); // consumes true

                    return Ok(Expression::BooleanLiteralExp {
                        name: "true".to_string(),
                        value: true,
                    });
                }

                KeywordKind::False => {
                    self.next(); // consumes false

                    return Ok(Expression::BooleanLiteralExp {
                        name: "false".to_string(),
                        value: false,
                    });
                }

                _ => {
                    return Err(format!(
                        "Given keyword does not have a prefix function {:?}",
                        keyword_kind
                    ))
                }
            },

            Token::Ident { name } => {
                // let name = format!("{}{}", name, context.suffix);

                if let Some(sym_meta) = context.get(&name) {
                    let suffix_name = format!("{}{}", name, context.suffix);

                    let exp = Ok(Expression::IdentExp {
                        name: suffix_name,
                        data_type: sym_meta.data_type.clone(),
                    });

                    self.next(); // Consumes ident

                    return exp;
                } else {
                    return Err(format!("There is no variable defined with name {}", name));
                }
            }

            Token::CurveOpenBracket => {
                self.next(); // consume (

                let grouped_exp = self.parse_expression(1, context)?;

                let cur_tok = self.get_cur_token().unwrap();

                if cur_tok == &Token::Eof {
                    println!("Contents : {:?}", self.content);
                }

                assert_eq!(cur_tok, &Token::CurveCloseBracket);
                self.next(); // consumes )

                return Ok(grouped_exp);
            }

            tok => {
                return Err(format!(
                    "Given token {:?} does not have not a prefix function",
                    tok
                ))
            }
        }
    }

    fn get_non_prefix_exp(
        &mut self,
        left: Expression,
        context: &SymbolContext,
    ) -> Result<Result<Expression, Expression>, String> {
        let non_prefix_token = self.get_cur_token().unwrap();

        match non_prefix_token {
            Token::Plus => {
                let precedence = Parser::get_non_prefix_precedence(&Token::Plus);

                self.next(); // consumes +

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::Plus,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::Minus => {
                let precedence = Parser::get_non_prefix_precedence(&Token::Minus);

                self.next(); // consumes -

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::Minus,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::Star => {
                let precedence = Parser::get_non_prefix_precedence(&Token::Star);

                self.next(); // consumes *

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::Star,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::Slash => {
                let precedence = Parser::get_non_prefix_precedence(&Token::Slash);

                self.next(); // consumes /

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::Slash,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::VerticalBar => {
                let precedence = Parser::get_non_prefix_precedence(&Token::VerticalBar);

                self.next(); // consumes |

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::VerticalBar,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::Caret => {
                let precedence = Parser::get_non_prefix_precedence(&Token::Caret);

                self.next(); // consumes ^

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::Caret,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::Ampersand => {
                let precedence = Parser::get_non_prefix_precedence(&Token::Ampersand);

                self.next(); // consumes &

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::Ampersand,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::StrictEquality => {
                let precedence = Parser::get_non_prefix_precedence(&Token::StrictEquality);

                self.next(); // consumes ===

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::StrictEquality,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::StrictNotEqual => {
                let precedence = Parser::get_non_prefix_precedence(&Token::StrictNotEqual);

                self.next(); // consumes !==

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::StrictNotEqual,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::LessThan => {
                let precedence = Parser::get_non_prefix_precedence(&Token::LessThan);

                self.next(); // consumes !==

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::LessThan,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::LessThanOrEqual => {
                let precedence = Parser::get_non_prefix_precedence(&Token::LessThanOrEqual);

                self.next(); // consumes !==

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::LessThanOrEqual,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::GreaterThan => {
                let precedence = Parser::get_non_prefix_precedence(&Token::GreaterThan);

                self.next(); // consumes !==

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::GreaterThan,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            Token::GreaterThanOrEqual => {
                let precedence = Parser::get_non_prefix_precedence(&Token::GreaterThanOrEqual);

                self.next(); // consumes !==

                let right_exp = Box::new(self.parse_expression(precedence, context)?);
                return Ok(Ok(Expression::BinaryExp {
                    operator: BinaryOperator::GreaterThanOrEqual,
                    left: Box::new(left),
                    right: right_exp,
                }));
            }

            _ => return Ok(Err(left)),
        }
    }

    fn get_prefix_precedence(token: &Token) -> usize {
        match token {
            Token::Plus | Token::Minus | Token::Bang => return 17,

            _ => return 1,
        }
    }

    fn get_non_prefix_precedence(token: &Token) -> usize {
        match token {
            Token::Star | Token::Slash => return 15,

            Token::Plus | Token::Minus => return 14,

            Token::LessThan
            | Token::LessThanOrEqual
            | Token::GreaterThan
            | Token::GreaterThanOrEqual => return 12,

            Token::StrictEquality | Token::StrictNotEqual => return 11,
            Token::Ampersand => return 10,
            Token::Caret => return 9,
            Token::VerticalBar => return 8,

            _ => return 1,
        }
    }

    fn parse_type_declaration(&mut self) -> DataType {
        let cur_tok = self.get_cur_token().unwrap();

        let data_type = match cur_tok {
            Token::Ident { name } => {
                if name == "string" {
                    DataType::String
                } else if name == "boolean" {
                    DataType::Boolean
                } else if name == "number" {
                    DataType::Float
                } else {
                    todo!()
                }
            }
            _ => todo!(),
        };

        self.next(); // consumes ident;

        return data_type;
    }
}
