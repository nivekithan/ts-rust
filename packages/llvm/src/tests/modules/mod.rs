use std::collections::HashMap;

use inkwell::context::Context;
use lexer::convert_to_token;
use parser::{parse_main, resolver::Resolver as ParserResolver};

use crate::{compile_parser_resolver_to_llvm_module, link_llvm_module_resolver};

#[test]
fn test_simple_import() {
    let main_file = "
    import { foo } from \"foo\";

    const y = foo(5)

    ";

    let foo_file = "

    export function foo(x : number) : number {
        return x + 5;
    };
    ";

    let mut dependent_files: HashMap<String, String> = HashMap::new();
    dependent_files.insert("foo".to_string(), foo_file.to_string());

    let mut parser_resolver = ParserResolver::from(dependent_files.clone());
    parse_main(convert_to_token(main_file), &mut parser_resolver);

    let context = Context::create();

    let llvm_resolver = compile_parser_resolver_to_llvm_module(parser_resolver, &context);
    let linked_module = link_llvm_module_resolver(llvm_resolver);

    let main_content = linked_module.get_string_representation().to_string();

    let input = format!(
        "
Main file:
{}

Foo File:
{}
",
        main_file, foo_file
    );

    insta::assert_snapshot!(input, main_content);
}

#[test]
fn test_complex_import() {
    let main_file = "
    import {foo} from \"foo\";
    import {boo} from \"boo\";

    const y = foo(5);
    const z = boo(y);
    ";

    let foo_file = "
    import {boo} from \"boo\";

    export function foo(x : number) : number {
        return boo(x) + 5;
    };
    ";

    let boo_file = "
    export function boo(x : number) : number {
        return 5*x;
    };
    ";

    let mut dependent_files: HashMap<String, String> = HashMap::new();
    dependent_files.insert("foo".to_string(), foo_file.to_string());
    dependent_files.insert("boo".to_string(), boo_file.to_string());
    let mut parser_resolver = ParserResolver::from(dependent_files.clone());
    parse_main(convert_to_token(main_file), &mut parser_resolver);

    let context = Context::create();

    let llvm_resolver = compile_parser_resolver_to_llvm_module(parser_resolver, &context);
    let linked_module = link_llvm_module_resolver(llvm_resolver);

    let main_content = linked_module.get_string_representation().to_string();

    //     let input = format!(
    // "
    // Main file:
    // {}

    // Foo File:
    // {}

    // Boo File:
    // {}
    // ", main_file, foo_file, boo_file);

    insta::assert_snapshot!(main_content);
}
