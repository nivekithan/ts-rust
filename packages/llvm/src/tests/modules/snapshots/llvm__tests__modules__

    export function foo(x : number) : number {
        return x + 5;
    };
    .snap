---
source: packages/llvm/src/tests/modules/mod.rs
expression: content

---
; ModuleID = 'foo'
source_filename = "foo"

define private void @main() {
entry:
  ret void
}

define double @foo_(double %0) {
main:
  %x_ = alloca double, align 8
  store double %0, double* %x_, align 8
  %"2" = load double, double* %x_, align 8
  %"1" = fadd double %"2", 5.000000e+00
  ret double %"1"
}

